File: Part 10 - Model Selection & Boosting/XGboost.py
# XGboost
# pip install xgboost

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv("Churn_Modelling.csv")
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categories=X[1], sparse=False)
X = onehotencoder.fit_transform(X)
X = X[:, 1:]


# Spliting dataset into trainning set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size=0.20,
                                                    random_state=0)

# Fitting XGboost to the Training set
from xgboost import XGBClassifier
classifier = XGBClassifier()
classifier.fit(X_train, y_train)


# Predicting the Test set results
y_pred = classifier.predict(X_test)


# Making the confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# Applying k-fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator=classifier, X=X_train,
                             y=y_train,
                             cv=10)
accuracies.mean()
accuracies.std()

File: Part 10 - Model Selection & Boosting/R/XGboost.R
# XGboost

# Importing the dataset
dataset = read.csv('Churn_Modelling.csv')
dataset = dataset[4:14]

# Encoding categorical data
dataset$Geography = as.numeric(factor(dataset$Geography,
																			levels = c('France', 'Spain', 'Germany'),
																			labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
																	 levels = c('Female', 'Male'),
																	 labels = c(1, 2)))


# Splitting dataset into training set and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.80)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Applying XGboost to the training set
# install.packages('xgboost')
library(xgboost)
classifier = xgboost(data = as.matrix(training_set[-11]),
										 label = training_set$Exited,
										 nrounds = 10)


# Applying k-fold CrossValidation
# install.packages('caret')
library(caret)
folds = createFolds(training_set$Exited, k = 10)
cv = lapply(folds, function(x) {
	training_fold = training_set[-x, ]
	test_fold = training_set[x, ]
	classifier = xgboost(data = as.matrix(training_set[-11]),
											 label = training_set$Exited,
											 nrounds = 10)
	y_pred = predict(classifier, newdata = as.matrix(test_fold[-11]) )
	y_pred = (y_pred >= 0.5)
	cm = table(test_fold[, 11], y_pred)
	accuracy = (cm[1, 1] + cm[2, 2]) / (cm[1, 1] + cm[1, 2] + cm[2, 1] + cm[2, 2])
	return(accuracy)
	
})
accuracy = mean(as.numeric(cv))
